src/queue.c

```diff
@@ -5693,6 +5694,11 @@ dispatch_main(void)
 		_dispatch_object_debug(&_dispatch_main_q, "%s", __func__);
 		_dispatch_program_is_probably_callback_driven = true;
 		_dispatch_ktrace0(ARIADNE_ENTER_DISPATCH_MAIN_CODE);
+#ifdef __linux__
+		pthread_key_t dispatch_main_key;
+		pthread_key_create(&dispatch_main_key, _dispatch_sig_thread);
+		pthread_setspecific(dispatch_main_key, &dispatch_main_key); // anything non NULL really
```

Comment:
the `// anything non null really` is a bad comment.
you should explain above that the point of creating that TSD key before pthread_exit() is because glibc will call the destructor of that TSD key last, which gets the thread parked in sigsuspend hopefully after all other TSD destructors have run (there are still holes as a TSD constructor could re-use a TSD and glibc has to loop over TSD keys to find out about this, and I'm pretty sure my hack prevents this, but that's as good it can be).